gson/src/main/java/com/google/gson/internal/bind/RecordTypeAdapterFactory.java

```diff
@@ -0,0 +1,480 @@
+            SerializedName serializedNameAnnotation =
+                    RECORD_HELPER.getAnnotation(recordComponent, SerializedName.class);
+            String serializedName;
+            String[] deserializedNames;
+            if (serializedNameAnnotation == null) {
+                serializedName = RECORD_HELPER.getName(recordComponent);
+                deserializedNames = new String[] { RECORD_HELPER.getName(recordComponent) };
+            }
+            else {
+                serializedName = serializedNameAnnotation.value();
+                deserializedNames = new String[serializedNameAnnotation.alternate().length + 1];
+                deserializedNames[0] = serializedName;
+                for (int i = 0; i < serializedNameAnnotation.alternate().length; i++) {
+                    deserializedNames[i + 1] = serializedNameAnnotation.alternate()[i];
+                }
+            }
+
+            return new RecordField(
+                    gson.serializeNulls(),
+                    excluder.excludeField(field, true),
+                    excluder.excludeField(field, false),
+                    RECORD_HELPER.getAccessor(recordComponent),
+                    serializedName,
+                    deserializedNames,
+                    typeAdapter
+            );
+        }
+    }
+
+    /**
+     * Helper class to contain all the rules for a single field, in regard to Gson annotations.
+     */
+    private static final class RecordField {
+        // Configured on Gson itself, if null values should be included in output
+        private final boolean serializeNulls;
+        // Configured via Expose / Since / Until annotations
+        private final boolean excludeOnSerialize;
+        private final boolean excludeOnDeSerialize;
+        // The public record method for reading out values from an instance of the record.
+        private final Method accessor;
+        // The name to write when serializing this field on a record
+        private final String serializedName;
+        // Which names we expect to find when deserializing this record
+        private final String[] deSerializedNames;
+        // Adapter for reading/write the json version of a record field.
+        private final TypeAdapter typeAdapter;
+
+        private RecordField(boolean serializeNulls, boolean excludeOnSerialize, boolean excludeOnDeSerialize,
+                Method accessor,
+                String serializedName, String[] deSerializedNames, TypeAdapter typeAdapter) {
+            this.serializeNulls = serializeNulls;
+            this.excludeOnSerialize = excludeOnSerialize;
+            this.excludeOnDeSerialize = excludeOnDeSerialize;
+            this.accessor = accessor;
+            this.serializedName = serializedName;
+            this.deSerializedNames = deSerializedNames;
+            this.typeAdapter = typeAdapter;
+        }
+
+        private void write(JsonWriter out, Object value) throws IOException {
```

Comment:
The names `write` and `read` suggest a kind of symmetry which is a bit misleading, since `write` writes both the name and the value and `read` reads just the value. Perhaps rename them `writeNameAndValue` and `readValue`?

Reply:
Done.